Expose WebCore::ContentExtensions::parseRuleList as WebKit SPI

b-weinstein opened the pull request:
#### a423cde6af09afbcb38784951a94ff40a28fac6e
<pre>
Expose WebCore::ContentExtensions::parseRuleList as WebKit SPI
<a href="https://bugs.webkit.org/show_bug.cgi?id=251465">https://bugs.webkit.org/show_bug.cgi?id=251465</a>
&lt;rdar://problem/99753935&gt;

Reviewed by Alex Christensen.

This patch also adds tests for passing and failing rule parsing.

* Source/WebKit/UIProcess/API/APIContentRuleList.cpp:
(API::ContentRuleList::parseRuleList): Call into WebCore::ContentExtensions::parseRuleList and return the error code if it exists.
* Source/WebKit/UIProcess/API/APIContentRuleList.h:
* Source/WebKit/UIProcess/API/Cocoa/WKContentRuleList.mm:
(+[WKContentRuleList _parseRuleList:]): Call into API::ContentRuleList::parseRuleList and wrap the error code in an NSError if necessary.
* Source/WebKit/UIProcess/API/Cocoa/WKContentRuleListPrivate.h:

Canonical link: <a href="https://commits.webkit.org/259706@main">https://commits.webkit.org/259706@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/51484f9ca5a3583ef865d52f445704a2761b3392

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105699 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14789 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38571 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114923 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/175064 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109600 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16196 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5986 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/97962 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114729 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111452 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12322 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/95302 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/97962 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/94186 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/26942 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/97962 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8055 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/28295 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8550 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/4881 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14171 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47842 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6708 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10103 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/API/APIContentRuleList.cpp:
```diff
@@ -77,6 +78,12 @@ bool ContentRuleList::supportsRegularExpression(const WTF::String& regex)
 
+std::error_code ContentRuleList::parseRuleList(const WTF::String& ruleList)
+{
+    Expected<Vector<WebCore::ContentExtensions::ContentExtensionRule>, std::error_code> result = WebCore::ContentExtensions::parseRuleList(ruleList);
```
auto

b-weinstein replied:
> This needs a unit test that causes a parse failure and a unit test that does not.

Yes it does. I'll add those in the next go of this PR.

achristensen07 commented on Source/WebKit/UIProcess/API/APIContentRuleList.cpp:
```diff
@@ -77,6 +78,12 @@ bool ContentRuleList::supportsRegularExpression(const WTF::String& regex)
+std::error_code ContentRuleList::parseRuleList(const WTF::String& ruleList)
+{
+    Expected<Vector<WebCore::ContentExtensions::ContentExtensionRule>, std::error_code> result = WebCore::ContentExtensions::parseRuleList(ruleList);
+    return result.error();
```
With Expected you have to check has_value() before calling error()

achristensen07 commented on Source/WebKit/UIProcess/API/Cocoa/WKContentRuleList.mm:
```diff
@@ -70,4 +71,18 @@ + (BOOL)_supportsRegularExpression:(NSString *)regex
+    if (!errorCode)
+        return nil;
+
+    return [NSError errorWithDomain:WKErrorDomain code:errorCode userInfo:nil];
```
This error code is not in WKErrorDomain.  Use WKErrorContentRuleListStoreCompileFailed and maybe put more info in the userInfo like we do in compileContentRuleListForIdentifier:

b-weinstein replied:
Sounds good - I see that example.

achristensen07 reviewed (changes requested):
This needs a unit test that causes a parse failure and a unit test that does not.

b-weinstein committed:
Expose WebCore::ContentExtensions::parseRuleList as WebKit SPI
https://bugs.webkit.org/show_bug.cgi?id=251465
<rdar://problem/99753935>

Reviewed by Alex Christensen.

This patch also adds tests for passing and failing rule parsing.

* Source/WebKit/UIProcess/API/APIContentRuleList.cpp:
(API::ContentRuleList::parseRuleList): Call into WebCore::ContentExtensions::parseRuleList and return the error code if it exists.
* Source/WebKit/UIProcess/API/APIContentRuleList.h:
* Source/WebKit/UIProcess/API/Cocoa/WKContentRuleList.mm:
(+[WKContentRuleList _parseRuleList:]): Call into API::ContentRuleList::parseRuleList and wrap the error code in an NSError if necessary.
* Source/WebKit/UIProcess/API/Cocoa/WKContentRuleListPrivate.h:

Canonical link: https://commits.webkit.org/259706@main
